samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,835 @@
+		rsp_send(evt->param.publish.message.topic.topic.utf8,
+			evt->param.publish.message.topic.topic.size);
+		rsp_send("\r\n", 2);
+		rsp_send(payload_buf,
+			evt->param.publish.message.payload.len);
+		rsp_send("\r\n", 2);
+	}
+}
+
+/**@brief MQTT client event handler
+ */
+void mqtt_evt_handler(struct mqtt_client *const c,
+		      const struct mqtt_evt *evt)
+{
+	int ret;
+
+	sprintf(rsp_buf, "#XMQTTEVT: %d,%d\r\n",
+		evt->type, evt->result);
+	rsp_send(rsp_buf, strlen(rsp_buf));
+
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT connect failed %d", evt->result);
+			break;
+		}
+		LOG_DBG("MQTT client connected!");
+		ctx.connected = true;
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT disconnect failed %d", evt->result);
+			break;
+		}
+		LOG_DBG("MQTT client disconnected %d\n", evt->result);
+		ctx.connected = false;
+		break;
+
+	case MQTT_EVT_PUBLISH:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT Publish event failed %d", evt->result);
+			break;
+		}
+		handle_mqtt_publish_evt(c, evt);
+		break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBACK error %d\n", evt->result);
+			break;
+		}
+		LOG_DBG("PUBACK packet id: %u\n",
+				evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_PUBREC:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBREC error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBREC packet id: %u", evt->param.pubrec.message_id);
+		{
+			struct mqtt_pubrel_param param = {
+				.message_id = evt->param.pubrel.message_id
+			};
+			ret = mqtt_publish_qos2_release(&client, &param);
+			if (ret) {
+				LOG_ERR("mqtt_publish_qos2_release: Failed! %d",
+					ret);
+			} else {
+				LOG_DBG("Release, id %u",
+					evt->param.pubrec.message_id);
+			}
+		}
+		break;
+
+	case MQTT_EVT_PUBREL:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBREL error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBREL packet id %u", evt->param.pubrel.message_id);
+		{
+			struct mqtt_pubcomp_param param = {
+				.message_id = evt->param.pubrel.message_id
+			};
+			ret = mqtt_publish_qos2_complete(&client, &param);
+			if (ret) {
+				LOG_ERR("mqtt_publish_qos2_complete Failed:%d",
+					ret);
+			} else {
+				LOG_DBG("Complete, id %u",
+					evt->param.pubrel.message_id);
+			}
+		}
+		break;
+
+	case MQTT_EVT_PUBCOMP:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBCOMP error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBCOMP packet id %u", evt->param.pubcomp.message_id);
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT SUBACK error %d\n", evt->result);
+			break;
+		}
+		LOG_DBG("SUBACK packet id: %u\n",
+				evt->param.suback.message_id);
+		break;
+
+	default:
+		LOG_DBG("default: %d\n", evt->type);
+		break;
+	}
+}
+
+static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+
+	while (1) {
+		if (client.internal.state < MQTT_STATE_TCP_CONNECTED) {
```

Comment:
The application should not use internal fields of the client structure (especiialy that it requires hacky solutions like copying the content of internal header). There's really no need for this.

The thread could be simply spawned during the initialization phase, and then we could have something like:
```
static K_SEM_DEFINE(socket_ready, 0, 1);
static bool mqtt_running;

...
static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3) {
	while (true) {
		k_sem_take(&socket_ready, K_FOREVER);

		while (mqtt_running) {
				/* Do all the processing. Any error reported by mqtt_ functions or poll events should break the loop. */
		}
	}
}
```

`do_mqtt_connect` shall call `k_sem_give(&socket_ready);` as it does so far plus set the `mqtt_running = true;`.
`do_mqtt_disconnect` shall set `mqtt_running = false;` to break the loop (verified also after poll() call).

Reply:
I've implement the above logic. ctx.connected is used instead of mqtt_running.